File: backend/entities/message.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

// Definimos el tipo MessageType como int
type MessageType int

// Ahora definimos constantes para los valores posibles de MessageType
const (
	// Los diferentes tipos de mensajes
	Text         MessageType = iota // 0
	Image                           // 1
	Video                           // 2
	Notification                    // 3
)

// KafkaMessage represents the structure of a message in Kafka
type KafkaMessage struct {
	Key       string                 `json:"key"`
	Value     string                 `json:"value"`
	Headers   map[string]interface{} `json:"headers"`
	Timestamp time.Time              `json:"timestap"`
}

// NatsMessage represents the structure of a message in NATS
type NatsMessage struct {
	Subject   string                 `json:"subject"`
	Data      []byte                 `json:"data"`
	Timestamp time.Time              `json:"timestap"`
	Headers   map[string]interface{} `json:"headers"`
}

// Clase Mensaje
type Metadata struct {
	AckStatus    bool   `json:"ackstatus"`    // Estado de confirmación
	Priority     int    `json:"priority"`     // Prioridad del mensaje
	OriginalLang string `json:"originallang"` // Idioma original
}
type Message struct {
	MessageId   uuid.UUID   `json:"messageid"`   // Identificador único
	MessageType MessageType `json:"messagetype"` // Tipo de mensaje
	SendDate    time.Time   `json:"senddate"`    // Fecha de envío
	ServerDate  time.Time   `json:"serverdate"`  // Fecha en el servidor
	Nickname    string      `json:"nickname"`    // Nombre del remitente
	Token       string      `json:"tokenjwt"`    // JWT del remitente
	MessageText string      `json:"messagetext"` // Contenido del mensaje
	RoomID      uuid.UUID   `json:"roomid"`      // Sala destino
	RoomName    string      `json:"roomname"`    // Nombre de la sala
	Metadata    Metadata    `json:"metadata"`    // Metadatos adicionales
}

/*
	1. priority: Indica la importancia del mensaje.
			Mensajes generales de chat: 1.
			Notificaciones importantes: 2.
			Alertas críticas: 3.
	2. originalLang: Representa el idioma original del mensaje. Utiliza códigos ISO 639-1 de dos letras para especificar el idioma.
			en: Inglés.
			es: Español.
			fr: Francés.
			de: Alemán.
			zh: Chino.
*/

// Convertir Priority si está presente
type RequestListUsers struct {
	RoomId      uuid.UUID `json:"roomid"`
	TokenSesion string    `json:"tokensesion"`
	Nickname    string    `json:"nickname"`
	Topic       string    `json:"topic"`
	X_GoChat    string    `json:"x_gochat"`
}
type RequestListMessages struct {
	Operation     string    `json:"operation"`
	LastMessageId uuid.UUID `json:"lastmessageid"`
	TokenSesion   string    `json:"tokensesion"`
	Nickname      string    `json:"nickname"`
	RoomId        uuid.UUID `json:"roomid"`
	Topic         string    `json:"topic"`
	X_GoChat      string    `json:"x_gochat"`
}
type AliveUsers struct {
	Nickname       string `json:"nickname"`
	LastActionTime string `json:"lastactiontime"`
}

type ResponseListUser struct {
	Status      string       `json:"status"`
	Message     string       `json:"message"`
	TokenSesion string       `json:"tokenSesion,omitempty"`
	Nickname    string       `json:"nickname,omitempty"`
	RoomId      uuid.UUID    `json:"roomId,omitempty"`
	X_GoChat    string       `json:"x_gochat,omitempty"`
	AliveUsers  []AliveUsers `json:"data,omitempty"`
}
type MessageResponse struct {
	MessageId   uuid.UUID `json:"messageid"`
	Nickname    string    `json:"nickname"`
	MessageText string    `json:"messagetext"`
}

// Estructura para la respuesta
type ResponseListMessages struct {
	Status      string            `json:"status"`
	Message     string            `json:"message"`
	TokenSesion string            `json:"tokenSesion"`
	Nickname    string            `json:"nickname"`
	RoomId      uuid.UUID         `json:"roomid"`
	X_GoChat    string            `json:"x_gochat"`
	ListMessage []MessageResponse `json:"data,omitempty"` // Lista de mensajes si existen
}

File: web_kafka/src/models/Message.ts
import { randomUUID } from 'crypto';
import { MessageData, Metadata, MessageType} from '../types/typesComm'; // Ajusta la ruta según la estructura de tu proyecto
import { validate as validateUUID } from 'uuid';
import { v4 as uuidv4 } from 'uuid'; 

export type UUID = string;
export class Message implements MessageData {
    messageId: string;      // Identificador único (UUID)
    messageType: MessageType; // Tipo de mensaje
    sendDate: string;       // Fecha de envío (ISO string)
    serverDate: string;     // Fecha en el servidor (ISO string)
    nickname: string;       // Nombre del remitente
    token: string;          // JWT del remitente
    messageText: string;    // Contenido del mensaje
    roomName: string;       // Nombre de la sala
    metadata: Metadata;     // Metadatos adicionales
  
    private _roomId: UUID    = "00000000-0000-0000-0000-000000000000"; 
    private _idMessage: UUID = "00000000-0000-0000-0000-000000000000"; 

    constructor(
        nickname: string,
        messageText: string,
        sendDate: string,
        roomId: string,
        token: string,
        roomName: string,
        metadata: Metadata,
        messageType: MessageType
    ) {
        this.nickname = nickname;
        this.messageText = messageText;
        this.sendDate = sendDate;
        this.serverDate = new Date().toISOString();  // Se asigna la fecha actual como fecha del servidor
        this.token = token;
        this.roomName = roomName;
        this.metadata = metadata;
        this.messageType = messageType;
        this.messageId =   uuidv4(); // Generación del UUID para el messageId
        this.roomId = roomId;        // Llamará al setter para validar el UUID
    }
 
    // Getter y Setter para idSala
    get idMessage(): UUID {
        return this.idMessage;
    }
    set idMessage(value: string) {
        if (!validateUUID(value)) {
            throw new Error('El ID de la sala no es un UUID válido');
        }
        this._idMessage = value; // Si la validación pasa, asigna el valor
    }
    // Getter y Setter para idSala
    get roomId(): UUID {
        return this._roomId;
    }
    set roomId(value: string) {
        if (!validateUUID(value)) {
            throw new Error('El ID de la sala no es un UUID válido');
        }
        this._roomId = value; // Si la validación pasa, asigna el valor
    }
    
  }
  // const msg = new Message('usuario1', 'Hola, ¿cómo estás?', new Date().toISOString(), '123e4567-e89b-12d3-a456-426614174000');
  // const msg = new Message('usuario2', '¡Hola!', new Date().toISOString(), 'invalid-uuid');
 